File: py/src/gameboy.py
import argparse

from src.cart import Cart
from src.cpu import CPU
from src.gpu import GPU
from src.clock import Clock
from src.buttons import Buttons
from src.ram import RAM


class GameBoy:
    def __init__(self, args: argparse.Namespace) -> None:
        self.cart = Cart(args.rom)
        self.ram = RAM(self.cart, debug=args.debug_ram)
        self.cpu = CPU(self.ram, debug=args.debug_cpu)
        self.gpu = GPU(self.cpu, debug=args.debug_gpu, headless=args.headless)
        self.buttons = Buttons(self.cpu, headless=args.headless)
        self.clock = Clock(self.buttons, args.frames, args.profile, args.turbo)

    def run(self) -> None:
        while True:
            self.tick()

    def tick(self) -> None:
        self.cpu.tick()
        self.gpu.tick()
        self.buttons.tick()
        self.clock.tick()

File: php/src/gameboy.php
<?php

require "apu.php";
require "cart.php";
require "cpu.php";
require "gpu.php";
require "clock.php";
require "buttons.php";
require "ram.php";


class GameBoy
{
    private Cart $cart;
    private RAM $ram;
    private CPU $cpu;
    private GPU $gpu;
    private APU $apu;
    private Buttons $buttons;
    private Clock $clock;

    public function __construct(Args $args)
    {
        $this->cart = new Cart($args->rom);
        $this->ram = new RAM($this->cart, $args->debug_ram);
        $this->cpu = new CPU($this->ram, $args->debug_cpu);
        $this->gpu = new GPU($this->cpu, $args->debug_gpu, $args->headless);
        $this->apu = new APU($args->silent, $args->debug_apu);
        $this->buttons = new Buttons($this->cpu, $args->headless);
        $this->clock = new Clock($this->buttons, $args->frames, $args->profile, $args->turbo);
    }

    public function run()
    {
        // This is an infinite loop on purpose
        // @phpstan-ignore-next-line
        while (true) {
            $this->tick();
        }
    }

    public function tick()
    {
        $this->cpu->tick();
        $this->gpu->tick();
        $this->buttons->tick();
        $this->clock->tick();
        $this->apu->tick();
    }
}
